tests/integration/helpers.py

```diff
@@ -521,6 +562,89 @@ def audit_log_line_sanity_check(entry) -> bool:
     return True
 
 
+async def get_unit_hostname(ops_test: OpsTest, unit_id: int, app: str) -> str:
+    """Get the hostname of a specific unit."""
+    _, hostname, _ = await ops_test.juju("ssh", f"{app}/{unit_id}", "hostname")
+    return hostname.strip()
+
+
+async def get_raw_application(ops_test: OpsTest, app: str) -> Dict[str, Any]:
+    """Get raw application details."""
+    ret_code, stdout, stderr = await ops_test.juju(
+        *f"status --model {ops_test.model.info.name} {app} --format=json".split()
+    )
+    if ret_code != 0:
+        logger.error(f"Invalid return [{ret_code=}]: {stderr=}")
+        raise Exception(f"[{ret_code=}] {stderr=}")
+    return json.loads(stdout)["applications"][app]
+
+
+async def get_application_units(ops_test: OpsTest, app: str) -> List[Unit]:
+    """Get fully detailed units of an application."""
+    # Juju incorrectly reports the IP addresses after the network is restored this is reported as a
+    # bug here: https://github.com/juju/python-libjuju/issues/738. Once this bug is resolved use of
+    # `get_unit_ip` should be replaced with `.public_address`
+    raw_app = await get_raw_application(ops_test, app)
+    units = []
+    for u_name, unit in raw_app["units"].items():
+        unit_id = int(u_name.split("/")[-1])
+        if not unit.get("address", False):
+            # unit not ready yet...
+            continue
+
+        unit = Unit(
+            id=unit_id,
+            name=u_name.replace("/", "-"),
+            ip=unit["address"],
+            hostname=await get_unit_hostname(ops_test, unit_id, app),
+            is_leader=unit.get("leader", False),
+            workload_status=Status(
+                value=unit["workload-status"]["current"],
+                since=unit["workload-status"]["since"],
+                message=unit["workload-status"].get("message"),
+            ),
+            agent_status=Status(
+                value=unit["juju-status"]["current"],
+                since=unit["juju-status"]["since"],
+            ),
+            app_status=Status(
+                value=raw_app["application-status"]["current"],
+                since=raw_app["application-status"]["since"],
+                message=raw_app["application-status"].get("message"),
+            ),
+        )
+
+        units.append(unit)
+
+    return units
+
+
+async def check_all_units_blocked_with_status(
+    ops_test: OpsTest, db_app_name: str, status: Optional[str]
+) -> None:
+    # this is necessary because ops_model.units does not update the unit statuses
+    for unit in await get_application_units(ops_test, db_app_name):
+        assert (
+            unit.workload_status.value == "blocked"
+        ), f"unit {unit.name} not in blocked state, in {unit.workload_status}"
+        if status:
+            assert (
+                unit.workload_status.message == status
+            ), f"unit {unit.name} not in blocked state, in {unit.workload_status}"
+
+
+async def wait_for_mongodb_units_blocked(
```

Comment:
Consider making this more generic for all statuses in a future PR. Refer to [`wait_until`](https://github.com/canonical/opensearch-operator/blob/main/tests/integration/helpers_deployments.py#L284)